### PR TITLE
Add a bunch of information to devicetree docs + remove <BOARD>.dts_compiled

### PR DESCRIPTION
**Pull request by ulfalizer**
Rewrite and add more information to the devicetree docs, e.g. to explain dtlib, edtlib, and `gen_defines.py`. Also add an example to give a feel for the format of the generated macros.

Also remove `<BOARD>.dts_compiled`, which is unused after the old devicetree scripts were removed.

Four commits:

```
doc: dts: Add more details, implementation notes, and examples

Rewrite most of the 'Input and output files' section to add more details
and implementation notes, and to make some description more more
concrete:

 - Mention dtlib, edtlib, and gen_defines.py, and explain what they do

 - Add an example of how macros get generated from the devicetree, and
   explain the DT_<node>_<property> format of the generated identifiers.

   Merge the 'Include files generation' section into the 'Input and
   output files' section at the same time.

 - Explain that the base devicetree and the overlays just get
   concatenated, and why this works

 - Add more details on how dts_fixup.h files work

 - Mention that the C dtc compiler is only run to catch errors and
   warnings from it

 - Mention that the concatenated devicetree appears in
   zephyr/<BOARD>.dts.pre.tmp

 - Mention /include/, which is the native DTS mechanism for including
   other files

 - Misc. other minor tweaks
```

```
dts: Remove generation of <BOARD>.dts_compiled

Unused after the old devicescript were removed in commit c8c35f76ab
("scripts: dts: Remove deprecated extract_dts_includes.py script"). The
old scripts relied on parsing the output of 'dts -Odts', which replaces
e.g. phandle references. The new scripts parse the DTS files directly.

The edit to doc/guides/build/build-config-phase.svg is to remove the box
with '*.dts_compiled' in it (and the arrows to/from it). https://draw.io
can be used to edit it.

Keep running the dtc compiler just to catch any warnings/errors from it.
```

```
doc: dts: Say <BOARD> instead of BOARD

Makes it clearer that it's not meant literally.
```

```
doc: dts: Use definition list for compatible/label/ref

See https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html.
Gives neater output compared to a bullet list.

Also tweak the text a bit.
```

Before for the last commit:

before

After:

after

### CODE REVIEW

**Review by carlescufi — CHANGES_REQUESTED** [review]
Great work. Just a couple of nitpicks.

File: doc/guides/dts/index.rst [hunk 1]
```diff
@@ -126,25 +126,30 @@ a gray background, except for the root node, which is shown using its path
...
-  SPI chip select line, or some other value depending on the kind of device the
-  node represents.
+compatible
+    Says what kind of device the node represents. The value is a
+    string in the format "vendor,device", like ``"avago,apds9960"``, or a
+    sequence of these, like ``"ti,hdc", "ti,hdc1010"``. The build system uses
+    the compatible property to find the right :ref:`binding <bindings>` for the
+    node.
+
+label
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/guides/dts/index.rst]: It's a property called 'label'. Do a

    git grep "label ="

to see a bunch of examples.

This text was just reformatted (and slightly tweaked). It's from commit 85fa523cf0c ("doc: dts: add more explanations, with diagrams").

File: doc/guides/dts/index.rst [hunk 2]
```diff
@@ -126,25 +126,30 @@ a gray background, except for the root node, which is shown using its path
...
+    The device's name according to Zephyr's :ref:`device_drivers`. The value
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/guides/dts/index.rst]: I think the label is more like an instance marker for identifying devices. `git grep "label ="` gives a bunch of examples.

There was a bunch of discussion earlier re. trying to get rid of `label`. This text is old at least. I just reformatted it into a definition list.

File: doc/guides/dts/index.rst [hunk 3]
```diff
@@ -126,25 +126,30 @@ a gray background, except for the root node, which is shown using its path
...
+    can be passed to :c:func:`device_get_binding()` to retrieve the
+    corresponding driver-level :ref:`struct device* <device_struct>`. This
+    pointer can then be passed to the correct driver API by application code to
+    interact with the device. For example, calling
+    ``device_get_binding("I2C_0")`` would return a pointer to a device
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/guides/dts/index.rst]: Note that it's a string. Still seems to be used like this.

There's `data->i2c = device_get_binding(config->i2c_name);` in a bunch of places for example. `i2c_name` is assigned e.g. as `.i2c_name = DT_INST_0_PANASONIC_AMG88XX_BUS_NAME`. That macro gets generated from a `label` property.

There's an `.i2c_port = CONFIG_CRYPTO_ATAES132A_I2C_PORT_NAME` too, where `CRYPTO_ATAES132A_I2C_PORT_NAME` defaults to "I2C_0".

File: doc/guides/dts/index.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (ulfalizer) [reply doc/guides/dts/index.rst]: edtlib doesn't do anything with `label` properties by the way (except expose them). All the special-casing is in `gen_defines.py`.

File: doc/guides/dts/index.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (carlescufi) [reply doc/guides/dts/index.rst]: Yes you are right, in the spec it is listed under "Miscellaneous properties" indeed

**Review by carlescufi — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**